packages/core/snapshot/manager.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/iotaledger/goshimmer/packages/core/tangleold"
+	"github.com/iotaledger/goshimmer/packages/node/database"
+	"github.com/iotaledger/hive.go/core/generics/event"
+	"github.com/iotaledger/hive.go/core/kvstore"
+)
+
+const (
+	prefixSolidEntryPoint byte = iota
+)
+
+// Manager is the snapshot manager.
+type Manager struct {
+	tangle          *tangleold.Tangle
+	notarizationMgr *notarization.Manager
+
+	baseStore kvstore.KVStore
+	sepsLock  sync.RWMutex
+}
+
+// NewManager creates and returns a new snapshot manager.
+func NewManager(store kvstore.KVStore, t *tangleold.Tangle, nmgr *notarization.Manager) (new *Manager) {
+	new = &Manager{
+		tangle:          t,
+		notarizationMgr: nmgr,
+	}
+
+	new.baseStore = store
+
+	new.tangle.ConfirmationOracle.Events().BlockAccepted.Attach(event.NewClosure(func(e *tangleold.BlockAcceptedEvent) {
+		e.Block.ForEachParent(func(parent tangleold.Parent) {
+			index := parent.ID.EpochIndex
+			if index < e.Block.EI() {
+				new.insertSolidEntryPoint(parent.ID)
+			}
+		})
+	}))
+
+	new.tangle.ConfirmationOracle.Events().BlockOrphaned.Attach(event.NewClosure(func(event *tangleold.BlockAcceptedEvent) {
+		new.removeSolidEntryPoint(event.Block, event.Block.LatestConfirmedEpoch())
+	}))
+
+	return
+}
+
+// CreateSnapshot creates a snapshot file from node with a given name.
+func (m *Manager) CreateSnapshot(snapshotFileName string) (header *ledger.SnapshotHeader, err error) {
+	ecRecord, lastConfirmedEpoch, err := m.tangle.Options.CommitmentFunc()
+	if err != nil {
+		return nil, err
+	}
+
+	// lock the entire ledger in notarization manager until the snapshot is created.
+	m.notarizationMgr.WriteLockLedger()
+	defer m.notarizationMgr.WriteUnlockLedger()
+
+	// lock the entire solid entry points storage until the snapshot is created.
+	m.sepsLock.Lock()
+	defer m.sepsLock.Unlock()
+
+	headerPord := func() (header *ledger.SnapshotHeader, err error) {
+		header = &ledger.SnapshotHeader{
+			FullEpochIndex: lastConfirmedEpoch,
+			DiffEpochIndex: ecRecord.EI(),
+			LatestECRecord: ecRecord,
+		}
+		return header, nil
+	}
+
+	sepsProd := NewSolidEntryPointsProducer(lastConfirmedEpoch, ecRecord.EI(), m)
+	outputWithMetadataProd := NewLedgerUTXOStatesProducer(lastConfirmedEpoch, m.notarizationMgr)
+	epochDiffsProd := NewEpochDiffsProducer(lastConfirmedEpoch, ecRecord.EI(), m.notarizationMgr)
+
+	header, err = CreateSnapshot(snapshotFileName, headerPord, sepsProd, outputWithMetadataProd, epochDiffsProd)
+
+	return
+}
+
+// LoadSolidEntryPoints add solid entry points to storage.
+func (m *Manager) LoadSolidEntryPoints(seps *SolidEntryPoints) {
+	if seps == nil {
+		return
+	}
+
+	for _, b := range seps.Seps {
+		m.insertSolidEntryPoint(b)
+	}
+}
+
+func (m *Manager) insertSolidEntryPoint(id tangleold.BlockID) error {
+	m.sepsLock.Lock()
+	defer m.sepsLock.Unlock()
+
+	sepsStore, err := m.baseStore.WithRealm(append([]byte{database.PrefixSnapshot, prefixSolidEntryPoint}, id.EpochIndex.Bytes()...))
```

Comment:
I've recently realized we have a more idiomatic way of doing this: `byteutils.ConcatBytes`